### PR TITLE
drivers: can: Add can support for stm32L4

### PR DESCRIPTION
**Pull request by alexanderwachter**
This pull request extends the stm32 can driver to support STM32L4 controllers.

### CODE REVIEW

File: dts/arm/st/stm32l432.dtsi [hunk 1]
```diff
@@ -5,3 +5,20 @@
  */
 
 #include <st/stm32l4.dtsi>
+#include <dt-bindings/clock/stm32_clock.h>
+
+/ {
+	soc {
+		can1: can@40006400 {
```
Comment (erwango) [inline dts/arm/st/stm32l432.dtsi]: If I'm correct, all STM32L4 SoC have this CAN instance and hence it could be put in stm32l4xx.dtsi

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/arm/st/stm32l432.dtsi (hunk 1 above)
Comment (alexanderwachter) [reply dts/arm/st/stm32l432.dtsi]: STM32L462CE does not have any CAN.
Should I move it anyway?

Comment (erwango) [reply dts/arm/st/stm32l432.dtsi]: ok, had miss this one and STM32L462RE
But, if I'm correct, this will be the only 2 without CAN support in L4 series.
Rather than adding it all other files, we can use `/delete-node/ &can_1;` in stm32l462re/ce.dtsi files.

Comment (karstenkoenig) [reply dts/arm/st/stm32l432.dtsi]: As far as I can see on this page all of the L4 series have CAN
http://www.st.com/en/microcontrollers/stm32l4-series.html?querycriteria=productId=SS1580

In particular both STM32L462CE and STM32L462RE have it listed on their page as well.

Comment (alexanderwachter) [reply dts/arm/st/stm32l432.dtsi]: @karstenkoenig I think you are right!

Comment (erwango) [reply dts/arm/st/stm32l432.dtsi]: Well, indeed, CAN is not mentioned as supported on the table while it is on the respective pages.
So anyway, let's move can node in stm32l4.dtsi and if required they can be deleted in stm32l462(c/r)e.dtsi files.
I'll try to get someone clarifying information on the website.

Comment (erwango) [reply dts/arm/st/stm32l432.dtsi]: @alexanderwachter , ok, thanks

File: dts/arm/st/stm32l4.dtsi [hunk 2]
```diff
@@ -170,6 +170,31 @@
 			status = "disabled";
 			label = "SPI_2";
 		};
+
+		usbotg_fs: otgfs@50000000 {
```
Comment (erwango) [inline dts/arm/st/stm32l4.dtsi]: This should be part of another PR

File: drivers/can/stm32_can.c [hunk 3]
```diff
@@ -43,18 +43,47 @@ static inline void can_stm32_get_msg_fifo(CAN_FIFOMailBox_TypeDef *mbox,
...
-	u32_t bus_off;
+	CAN_FIFOMailBox_TypeDef *mbox;
+	int filter_match_index;
+	struct can_msg msg;
 
-	dev = (struct device *)arg;
-	data = DEV_DATA(dev);
-	cfg = DEV_CFG(dev);
-	can = cfg->can;
+	while (can->RF0R & CAN_RF0R_FMP0_Msk) {
```
Comment (erwango) [inline drivers/can/stm32_can.c]: We usually don't use *_Msk defines.
(One point being the "w/o _Msk define" is commented with bit signification, while the _Msk one is commented with value)
It seems this was present in initial PR, but I did not catch it.
Could you fix this ?

File: dts/arm/st/stm32l4.dtsi [hunk 4]
```diff
@@ -170,6 +170,31 @@
...
+			label= "OTGFS";
+		};
+
+		can1: can@40006400 {
+			compatible = "st,stm32-can";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40006400 0x400>;
+			interrupts = <19 0>, <20 0>, <21 0>, <22 0>;
+			interrupt-names = "TX", "RX0", "RX1", "SCE";
```
Comment (erwango) [inline dts/arm/st/stm32l4.dtsi]: If interrupt-names is required , you should add it in yaml binding (as optional since f0 do not use it).
It seems it worked even though it should be present in your case, so generation script would require a fix (or not).
Anyway, please fix the binding.

**Review by erwango — CHANGES_REQUESTED** [review]
Minor comments, otherwise looks good

File: drivers/can/stm32_can.c (hunk 3 above)
Comment (alexanderwachter) [reply drivers/can/stm32_can.c]: I will fix this in my next commit which is a rework of the PR before.
In this PR the bit timing is moved to device tree so that we could support controllers with more than one CAN. Additionally some checks for prescaler and issue #7933 is fixed there.

File: dts/arm/st/stm32l4.dtsi (hunk 4 above)
Comment (alexanderwachter) [reply dts/arm/st/stm32l4.dtsi]: Done. The unused bindings of port and pin will be removed in the rework PR

**Review by erwango — APPROVED** [review]
Change approved.

In a next PR, can you consider removing  rx-pin and  tx-pin from yaml binding? They seem unused.

**Comment by alexanderwachter** [comment]
@erwango I planed to remove it in the last PR but I somehow missed it. But there will be an API extension soon. I'll include it there.

**Comment by alexanderwachter** [comment]
@nashif  Done

**Comment by erwango** [comment]
@aurel32 : 
Thanks for this remark. So for now, I think it is ok to keep CAN statu ok in device tree, since actual feature activation is controlled by Kconfig CAN symbol, which should be deactivated by default.

Then, indeed, probably applying #6858, nucleo_l432kc pin configuration would be defaulted to arduino nano and then PA11 mapped to SPI1 CS, which makes current CAN pin configuration (PA11/PA12) conflicting.
In that case can overlay should also provide pin configuration (once dts pin configuration is effective).
